### ISSUE TITLE
Distillation : compresser un grand modèle

### ISSUE THREAD

**Issue by khalilbenaz** [khalilbenaz/khalilbenaz.github.io#479]
---
lede: La distillation transfère le savoir d'un grand modèle vers un petit, offrant 90 % des performances pour une fraction du coût d'inférence — à condition de bien choisir ce qu'on transfère.
lede_en: Distillation transfers knowledge from a large model into a small one, delivering 90% of the performance at a fraction of the inference cost — provided you choose what to transfer well.
title_en: Distillation: compressing a large model
---

## Le problème : le grand modèle est trop cher à servir

Un modèle de 400 milliards de paramètres raisonne merveilleusement bien et coûte une fortune à chaque requête. En production, on veut la qualité du grand et le prix du petit. La **distillation** promet exactement ça : entraîner un petit modèle (l'*élève*) à imiter un grand (le *professeur*), pour capturer l'essentiel du comportement dans une enveloppe dix fois plus légère.

Le pari n'est pas absurde. Un grand modèle est souvent sur-paramétré pour une tâche donnée ; sa capacité sert à *tout* savoir, pas à bien faire *une* chose. Si votre application couvre un domaine restreint, un élève spécialisé peut égaler le professeur sur ce domaine précis tout en étant bien plus rapide.

- **Objectif** — 90 % de la qualité, ~10 % du coût d'inférence.
- **Cible idéale** — tâche ou domaine borné, pas « intelligence générale ».
- **Mauvaise cible** — vouloir un petit modèle aussi polyvalent qu'un grand : ça ne tient pas.

## Comment on distille

Trois grandes familles, souvent combinées :

- **Distillation de logits (classique)** — l'élève apprend non pas la bonne réponse (label dur) mais **la distribution de probabilités** du professeur (labels doux). Ces probabilités portent l'information « le chat était le 2e choix le plus probable », que le label dur écrase.
- **Distillation par données** — le professeur génère un grand corpus de réponses, l'élève est fine-tuné dessus. C'est la forme la plus répandue avec les LLM, car elle ne demande pas d'accès aux logits internes.
- **Distillation de raisonnement** — le professeur produit des chaînes de raisonnement complètes, l'élève apprend à les reproduire. Elle transfère le *comment*, pas seulement le *quoi*.

```python
# Perte de distillation classique (labels doux)
import torch.nn.functional as F

def distill_loss(student_logits, teacher_logits, T=2.0):
    p_teacher = F.softmax(teacher_logits / T, dim=-1)  # adoucir
    log_p_student = F.log_softmax(student_logits / T, dim=-1)
    return F.kl_div(log_p_student, p_teacher, reduction="batchmean") * (T * T)
```

La température `T` élargit la distribution : plus elle est haute, plus les « choix secondaires » du professeur pèsent dans l'apprentissage.

## Distillation vs quantization vs pruning

On confond souvent ces trois voies de compression. Elles sont complémentaires, pas concurrentes :

- **Distillation** — change *l'architecture* : nouveau modèle plus petit, ré-entraîné. Effort élevé, gain de qualité/taille le meilleur.
- **Quantization** — garde le modèle, réduit la *précision* des poids (16→4 bits). Effort faible, quasi gratuit, dégradation limitée.
- **Pruning** — retire les poids/neurones peu utiles. Efficace en théorie, capricieux en pratique sur les transformers.

En 2026, le combo gagnant est souvent : **distiller** vers un modèle plus petit, *puis* **quantizer** l'élève. On empile les gains.

## Les pièges

La distillation échoue silencieusement quand on ne surveille pas :

- **Effondrement de couverture** — l'élève apprend les cas fréquents et rate les rares. Il paraît excellent sur le jeu de test moyen et médiocre sur la longue traîne.
- **Héritage des erreurs** — l'élève ne peut pas dépasser le professeur ; il hérite aussi de ses hallucinations et biais.
- **Domaine trop large** — vouloir distiller « toutes les capacités » dilue le transfert. Plus le domaine est étroit, meilleure est la distillation.
- **Fuite du jeu de test** — si le professeur a généré les données d'entraînement *et* de test, les métriques sont gonflées.

## Verdict

La distillation est l'outil de compression le plus puissant quand on accepte sa règle d'or : **elle marche d'autant mieux que la tâche est bornée**. Pour un classifieur, un extracteur, un assistant de domaine — excellente, avec un ratio qualité/coût imbattable. Pour reproduire une intelligence généraliste dans un petit modèle — illusoire, la capacité manque physiquement. La bonne démarche en 2026 : identifier précisément ce que votre application a *vraiment* besoin de savoir, distiller là-dessus, mesurer la longue traîne (pas seulement la moyenne), puis quantizer par-dessus. Un petit modèle bien distillé sur son domaine bat un grand modèle générique sur le coût — et souvent sur la latence perçue.

## The problem: the big model is too expensive to serve

A 400-billion-parameter model reasons wonderfully and costs a fortune per request. In production, you want the big one's quality and the small one's price. **Distillation** promises exactly that: train a small model (the *student*) to imitate a large one (the *teacher*), to capture the essence of the behavior in an envelope ten times lighter.

The bet isn't absurd. A large model is often over-parameterized for a given task; its capacity serves to know *everything*, not to do *one* thing well. If your application covers a narrow domain, a specialized student can match the teacher on that precise domain while being far faster.

- **Goal** — 90% of the quality, ~10% of the inference cost.
- **Ideal target** — a bounded task or domain, not "general intelligence."
- **Bad target** — wanting a small model as versatile as a large one: it won't hold.

## How you distill

Three main families, often combined:

- **Logit distillation (classic)** — the student learns not the right answer (hard label) but **the teacher's probability distribution** (soft labels). Those probabilities carry the "cat was the 2nd most likely choice" information that the hard label crushes.
- **Data distillation** — the teacher generates a large corpus of answers, the student is fine-tuned on it. This is the most common form with LLMs, since it doesn't require access to internal logits.
- **Reasoning distillation** — the teacher produces full reasoning chains, the student learns to reproduce them. It transfers the *how*, not just the *what*.

```python
# Classic distillation loss (soft labels)
import torch.nn.functional as F

def distill_loss(student_logits, teacher_logits, T=2.0):
    p_teacher = F.softmax(teacher_logits / T, dim=-1)  # soften
    log_p_student = F.log_softmax(student_logits / T, dim=-1)
    return F.kl_div(log_p_student, p_teacher, reduction="batchmean") * (T * T)
```

The temperature `T` widens the distribution: the higher it is, the more the teacher's "secondary choices" weigh in learning.

## Distillation vs quantization vs pruning

These three compression paths are often confused. They're complementary, not competing:

- **Distillation** — changes the *architecture*: a new smaller model, retrained. High effort, best quality/size gain.
- **Quantization** — keeps the model, reduces weight *precision* (16→4 bits). Low effort, nearly free, limited degradation.
- **Pruning** — removes low-usefulness weights/neurons. Effective in theory, finicky in practice on transformers.

In 2026, the winning combo is often: **distill** to a smaller model, *then* **quantize** the student. You stack the gains.

## The pitfalls

Distillation fails silently when you don't watch it:

- **Coverage collapse** — the student learns frequent cases and misses rare ones. It looks excellent on the average test set and mediocre on the long tail.
- **Error inheritance** — the student can't exceed the teacher; it also inherits its hallucinations and biases.
- **Domain too broad** — trying to distill "all capabilities" dilutes the transfer. The narrower the domain, the better the distillation.
- **Test set leakage** — if the teacher generated both the training *and* test data, metrics are inflated.

## Verdict

Distillation is the most powerful compression tool when you accept its golden rule: **it works better the more bounded the task is**. For a classifier, an extractor, a domain assistant — excellent, with an unbeatable quality/cost ratio. For reproducing generalist intelligence in a small model — illusory, the capacity physically isn't there. The right approach in 2026: pinpoint precisely what your application *truly* needs to know, distill on that, measure the long tail (not just the average), then quantize on top. A small model well distilled on its domain beats a large generic model on cost — and often on perceived latency.
